### PR TITLE
feat(cli): GB-4358: introduce a more thorough set of arguments to the dev command

### PR DESCRIPTION
**Pull request by jakubadamw**
# Description

These will control the level of logs for GraphQL operations and functions (resolvers, authorizers) etc.

```
Run your Grafbase project locally
Usage: grafbase dev [OPTIONS]
Options:
  -p, --port <PORT>                                                  Use a specific port [default: 4000]
  -s, --search                                                       If a given port is unavailable, search for another
      --disable-watch                                                Do not listen for schema changes and reload
      --log-level-functions <LOG_LEVEL_FUNCTIONS>                    Log level to print from function invocations, defaults to or overrides ‘log-level’ [possible values: none, error, warn, info, debug]
      --log-level-graphql-operations <LOG_LEVEL_GRAPHQL_OPERATIONS>  Log level to print for GraphQL operations, defaults to or overrides ‘log-level’ [possible values: none, error, warn, info, debug]
      --log-level <LOG_LEVEL>                                        Default log level to print [default: info] [possible values: none, error, warn, info, debug]
  -h, --help                                                         Print help
```

# Type of change

- [ ] 💔 Breaking
- [X] 🚀 Feature
- [ ] 🐛 Fix
- [ ] 🛠️ Tooling
- [ ] 🔨 Refactoring
- [ ] 🧪 Test
- [ ] 📦 Dependency
- [ ] 📖 Requires documentation update

### CODE REVIEW

File: cli/crates/cli/src/cli_input.rs [hunk 1]
```diff
@@ -17,9 +48,24 @@ pub struct DevCommand {
     /// Do not listen for schema changes and reload
     #[arg(long)]
     pub disable_watch: bool,
-    /// Log more events taking place such as introspection queries
-    #[arg(short, long)]
-    pub debug: bool,
+    /// Log level to print from function invocations, defaults to 'log-level'
+    #[arg(long)]
```
Comment (yoav-lavi) [inline cli/crates/cli/src/cli_input.rs:52]: I'd maybe name this since the parameter would be "FUNCTION_LOG_LEVEL"

Comment (yoav-lavi) [reply cli/crates/cli/src/cli_input.rs:52]: The others as well

File: cli/crates/cli/src/output/report.rs [hunk 2]
```diff
@@ -140,16 +143,68 @@ pub fn complete_udf_build(udf_kind: UdfKind, udf_name: &str, duration: std::time
...
+        LogEventType::OperationStarted { .. } => return,
+        LogEventType::OperationCompleted { name, duration, r#type } => {
+            let colour = match r#type {
+                common::types::OperationType::Query { is_introspection } => {
+                    if is_introspection && log_level_filter < LogLevel::Debug {
+                        return;
+                    }
+                    watercolor::colored::Color::Green
+                }
+                // Pink.
```
Comment (yoav-lavi) [inline cli/crates/cli/src/output/report.rs:187]: WDYM pink?

File: cli/crates/cli/src/cli_input.rs (hunk 1 above)
Comment (jakubadamw) [reply cli/crates/cli/src/cli_input.rs:52]: Sorry, not following. What parameter would be 'FUNCTION_LOG_LEVEL'?

File: cli/crates/cli/src/output/report.rs (hunk 2 above)
Comment (jakubadamw) [reply cli/crates/cli/src/output/report.rs:187]: It used to be pink, now I changed it to green, but didn't update the comment… The danger of comments.

File: cli/crates/cli/src/dev.rs [hunk 3]
```diff
@@ -16,7 +17,13 @@ static READY: Once = Once::new();
 ///
 /// returns [`CliError::ServerPanic`] if the development server panics
 #[allow(clippy::fn_params_excessive_bools)]
-pub fn dev(search: bool, watch: bool, external_port: u16, debug: bool, tracing: bool) -> Result<(), CliError> {
+pub fn dev(
```
Comment (pimeys) [inline cli/crates/cli/src/dev.rs:20]: Not part of this PR, but I'd combine all the boolean flags into named bitflags.

File: cli/crates/cli/src/cli_input.rs (hunk 1 above)
Comment (pimeys) [reply cli/crates/cli/src/cli_input.rs:52]: @yoav-lavi can you elaborate?

File: cli/crates/cli/src/output/report.rs [hunk 4]
```diff
@@ -140,16 +143,65 @@ pub fn complete_udf_build(udf_kind: UdfKind, udf_name: &str, duration: std::time
...
+        LogLevel::Debug => watercolor::colored::Color::BrightBlack,
+        LogLevel::Error => watercolor::colored::Color::Red,
+        LogLevel::Info => watercolor::colored::Color::Cyan,
+        LogLevel::Warn => watercolor::colored::Color::Yellow,
+    };
+    println!("{}", format!("[{udf_kind} '{udf_name}'] {message}").color(colour));
 }
 
-pub fn operation_started(_request_id: &str, _name: &Option<String>) {}
+pub fn operation_log(_request_id: &str, log_event_type: LogEventType, log_level_filter: Option<LogLevel>) {
```
Comment (pimeys) [inline cli/crates/cli/src/output/report.rs:169]: Do we need to have the request id parameter here if we don't use it?

**Review by pimeys — COMMENTED** [review]
Just a few nitpicks. In general I wonder could this be handled with tracing and writing a custom subscriber? I know it sucks a bit writing subscribers for tracing, and also it might not be feasible with our wasm setup. I think the dev command starts to have a bit too many parameters already.

@yoav-lavi should give his final blessing, he still has an open question.

File: cli/crates/server/src/types.rs [hunk 5]
```diff
@@ -23,18 +42,12 @@ pub enum ServerMessage {
...
-        request_id: String,
-        name: Option<String>,
-    },
-    OperationCompleted {
+    OperationLogMessage {
         request_id: String,
-        name: Option<String>,
-        duration: std::time::Duration,
-        r#type: common::types::OperationType,
+        event_type: LogEventType,
```
Comment (Finistere) [inline cli/crates/server/src/types.rs:50]: Maybe wrap `LogEvent` here? or is there a difference?

File: cli/crates/cli/src/output/report.rs [hunk 6]
```diff
@@ -140,16 +143,65 @@ pub fn complete_udf_build(udf_kind: UdfKind, udf_name: &str, duration: std::time
...
+    let Some(log_level_filter) = log_level_filter else {
+        return;
+    };
+    if log_level_filter < LogLevel::Info {
```
Comment (Finistere) [inline cli/crates/cli/src/output/report.rs:173]: if log level is debug you don't log anything? 🤔

File: cli/crates/cli/src/cli_input.rs [hunk 7]
```diff
@@ -1,11 +1,42 @@
...
+pub enum LogLevelFilter {
+    None,
+    Error,
+    Warn,
+    Info,
+    Debug,
+}
+
+impl From<LogLevelFilter> for Option<LogLevel> {
+    fn from(value: LogLevelFilter) -> Self {
```
Comment (Finistere) [inline cli/crates/cli/src/cli_input.rs:23]: Not sure how that would work out, but I wonder if this could just be done with something like `impl PartialOrd<LogLevel> for LogLevelFilter` or something like that

**Review by Finistere — APPROVED** [review]

File: cli/crates/cli/src/output/report.rs (hunk 4 above)
Comment (jakubadamw) [reply cli/crates/cli/src/output/report.rs:169]: Right now we don't, but in the future we most likely will

Comment (jakubadamw) [reply cli/crates/cli/src/output/report.rs:169]: I can remove it, though. Sure!

File: cli/crates/cli/src/output/report.rs (hunk 6 above)
Comment (jakubadamw) [reply cli/crates/cli/src/output/report.rs:173]:
```
#[derive(Clone, Copy, Debug, PartialEq, PartialOrd, serde::Deserialize)]
pub enum LogLevel {
    Error,
    Warn,
    Info,
    Debug,
}
```

🙂

File: cli/crates/cli/src/cli_input.rs (hunk 1 above)
Comment (yoav-lavi) [reply cli/crates/cli/src/cli_input.rs:52]: Specifically referring to ` --log-level-functions <LOG_LEVEL_FUNCTIONS>`, `LOG_LEVEL_FUNCTIONS` as a parameter name seems a bit weirdly put

**Comment by jakubadamw** [comment]
> Just a few nitpicks. In general I wonder could this be handled with tracing and writing a custom subscriber? I know it sucks a bit writing subscribers for tracing, and also it might not be feasible with our wasm setup. I think the dev command starts to have a bit too many parameters already.
> 
> @yoav-lavi should give his final blessing, he still has an open question.

Well, but it's a UX/UI design choice to have those parameters rather than a single 'RUST_LOG'-like parameter/environment variable. As for using tracing for this – it may be plausible, yes. I can look into refactoring this when I am back.

File: cli/crates/cli/src/cli_input.rs (hunk 7 above)
Comment (jakubadamw) [reply cli/crates/cli/src/cli_input.rs:23]: I considered that but to me operations like `PartialOrd` etc. should be defined between types that are equivalent. Semantically, these aren't, actually. `LogLevel` is the level of a single message, whereas filter belongs to another category (includes `None` that a level can't be). Doesn't feel like `PartialOrd` would be appropriate…

Comment (Finistere) [reply cli/crates/cli/src/cli_input.rs:23]: Well don't have a strong opinion, I think it probably would have made code simpler rather than dealing with the optional special case. 🤷

File: cli/crates/server/src/types.rs (hunk 5 above)
Comment (jakubadamw) [reply cli/crates/server/src/types.rs:50]: One is the type for the bridge protocol. I guess there isn't a difference, but we have separate types for the other kinds of things happening (UDF invocation) to stress the separation of these protocols. I can take a look at merging these, but we should do it for all in that case.

Comment (Finistere) [reply cli/crates/server/src/types.rs:50]: fair enough!
